File: 백준/Silver/1629. 곱셈/곱셈.py
a, b, c = map(int, input().split())

def remain(n):

    if n == 1:
        return a % c
    else:
        if n % 2 == 0:
            return (remain(n // 2) ** 2) % c
        else:
            return (remain(n // 2 + 1) * remain(n // 2)) % c

print(remain(b))

File: 백준/Silver/1629. 곱셈/곱셈.java


import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	static long A, B, C;

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		A = Integer.parseInt(st.nextToken());
		B = Integer.parseInt(st.nextToken());
		C = Integer.parseInt(st.nextToken());

		System.out.println(remain(B));
	}

	private static long remain(long n) {
		if (n == 1) {
			return A % C;
		}

		long temp = remain(n / 2);

		if (n % 2 == 0) {
			return temp * temp % C;
		} else {
			return (temp * temp % C) * A % C;
		}
	}

}
